### PR TITLE
DrivAerML: no-Fourier ablation + gc=1.0 — are Fourier features essential?

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Every DrivAerML experiment uses `--enable-fourier`. This ablation tests whether Fourier features are actually contributing to the 4.619% result on DrivAerML, or whether they are a holdover from AirfRANS/TandemFoil where they were first validated. On AirfRANS: dropping Fourier features (PR #2455) tripled epoch count per unit time (~3x faster epochs) and still achieved reasonable results. On DrivAerML, each epoch is ~0.67 min at 4L/512d with Fourier. Without Fourier, epochs could be faster, allowing ~400-500 total epochs in the same 360-min budget. If Fourier is not essential for the 3D geometric encoding, removing it could actually improve results via higher epoch count.

## Instructions

**IMPORTANT: Use EXACTLY these run specifications.**

Run 1 — DrivAerML no-Fourier + gc=1.0 (ablation):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --no-use-ema \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group levi-fourier-ablation
```

Run 2 — DrivAerML no-Fourier + gc=1.0 + WD=1e-2 (compound without Fourier):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 5e-4 --cosine-t-max 30 --no-use-ema \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group levi-fourier-ablation
```

Run 3 — DrivAerML no-Fourier + lr=7e-4 + gc=1.0 + WD=1e-2 (higher LR without Fourier):
```bash
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw \
  --lr 7e-4 --cosine-t-max 30 --no-use-ema \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb-group levi-fourier-ablation
```

Run 4 — AirfRANS no-Fourier + gc=1.0 + WD=1e-2 at 2L/256d (cross-dataset, mirroring AirfRANS finding):
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 7e-4 --cosine-t-max 5 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 \
  --wandb-group levi-fourier-ablation
```

Report `val_primary/surface_rel_l2_pct` for DrivAerML and the per-epoch training speed (minutes/epoch) for Runs 1-3. Compare to baseline golden ~0.67 min/epoch. `val_primary/surface_mse` for AirfRANS.

## Baseline

- **DrivAerML val_primary/surface_rel_l2_pct:** 4.619% (PR #2691, epoch 256, WITH Fourier)
- **AirfRANS val_primary/surface_mse:** 0.001236 (PR #2828, epoch 349)
- **DrivAerML external target:** 3.71% (AB-UPT)
- **Reproduce golden DrivAerML (WITH Fourier):** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT levi:

## Results

### Summary

**Fourier features are essential for DrivAerML and AirfRANS.** Removing them causes catastrophic performance degradation on both datasets, with most runs eventually diverging. The no-Fourier models cannot recover even with more epochs, higher LR, weight decay, or gradient clipping.

### Run 1 — DrivAerML no-Fourier + gc=1.0

| Metric | Value |
|---|---|
| **val_primary/surface_rel_l2_pct (best)** | **59.63%** (ep ~440) |
| val_primary/surface_rel_l2_pct (final, ep457) | 59.67% |
| test_primary/surface_rel_l2_pct | 60.24% |
| **Baseline (WITH Fourier)** | **4.619%** (ep256) |
| **External target** | **3.71%** (AB-UPT) |
| **Degradation** | **12.9x worse** |
| Epochs | 457 (~0.79 min/epoch) |
| Peak grad_norm_mean | Infinity (all 394 batches clipped every epoch) |
| W&B run | `zfab2jda` |

**Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent levi --wandb-name "levi/no-fourier-gc1.0" --wandb-group levi-fourier-ablation`

**Analysis:** The model flat-lined at ~59.6% from early training and never improved. Gradient norms were Infinity throughout, meaning gc=1.0 was constantly clipping but the model had no useful gradient signal. Without Fourier features, the model lacks the frequency-domain encoding needed to resolve DrivAerML's 3D surface pressure patterns.

---

### Run 2 — DrivAerML no-Fourier + gc=1.0 + WD=1e-2

| Metric | Value |
|---|---|
| **val_primary/surface_rel_l2_pct (best)** | **11.03%** (ep54) |
| val_primary/surface_rel_l2_pct (final) | NaN (diverged at ~ep155) |
| test_primary/surface_rel_l2_pct | NaN |
| **Baseline (WITH Fourier)** | **4.619%** |
| **Degradation at best** | **2.4x worse** |
| Epochs (valid) | 154 of 458 total |
| W&B run | `sdps35r7` |

**Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 1e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent levi --wandb-name "levi/no-fourier-gc1.0-wd1e2" --wandb-group levi-fourier-ablation`

**Analysis:** WD=1e-2 provided enough regularization for the model to initially learn (11.03% at ep54), but the optimization was unstable — the model oscillated wildly between 46-79% in later epochs before fully diverging to NaN at ~ep155. WD helped early training but couldn't compensate for the missing Fourier encoding.

---

### Run 3 — DrivAerML no-Fourier + lr=7e-4 + gc=1.0 + WD=1e-2

| Metric | Value |
|---|---|
| **val_primary/surface_rel_l2_pct (best)** | **6.78%** (ep153) |
| val_primary/surface_rel_l2_pct (final, ep457) | 81.63% |
| test_primary/surface_rel_l2_pct | 85.12% |
| **Baseline (WITH Fourier)** | **4.619%** |
| **Degradation at best** | **1.47x worse** |
| Epochs | 457 (~0.79 min/epoch) |
| W&B run | `3co9ll3u` |

**Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py --dataset drivaerml --optimizer adamw --lr 7e-4 --cosine-t-max 30 --no-use-ema --model-layers 4 --model-hidden-dim 512 --model-heads 8 --grad-clip 1.0 --weight-decay 1e-2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent levi --wandb-name "levi/no-fourier-lr7e4-gc1.0-wd1e2" --wandb-group levi-fourier-ablation`

**Analysis:** The higher LR + WD combination achieved the best no-Fourier result (6.78% at ep153), but this was a transient spike — the model then degraded monotonically to 81.63% by ep457. Gradient norms went to Infinity from ~ep153 onward. The lr=7e-4 variant found a briefly deeper basin but couldn't sustain it. Even at its best, 6.78% is still 47% worse than the Fourier baseline (4.619%).

---

### Run 4 — AirfRANS no-Fourier + gc=1.0 + WD=1e-2 at 2L/256d

| Metric | Value |
|---|---|
| **val_primary/surface_mse (best)** | **0.005615** (ep134) |
| val_primary/surface_mse (final) | NaN (diverged at ~ep168) |
| **Baseline (WITH Fourier 2L/256d)** | **0.001236** (ep349) |
| **External target** | **0.0043** |
| **Degradation at best** | **4.5x worse** |
| Epochs (valid) | 167 of 823 total |
| W&B run | `l4tr3jsv` |

**Command:** `cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent levi --wandb-name "levi/airfrans-no-fourier-gc1.0-wd1e2" --wandb-group levi-fourier-ablation`

**Analysis:** No-Fourier AirfRANS achieved 0.005615 at best (4.5x worse than baseline 0.001236), then diverged to NaN at ep168. The remaining 655 epochs produced all NaN. This contradicts the PR #2455 finding where no-Fourier AirfRANS got 0.2597 in 6 epochs — the current result with gc+WD did reach a much better 0.005615, but the 2L/256d architecture without Fourier is fundamentally unstable and cannot match the Fourier baseline. Fourier is critical for AirfRANS as well, though less catastrophically than DrivAerML.

---

### Per-epoch speed comparison

| Run | min/epoch | Baseline (WITH Fourier) | Speedup |
|---|---|---|---|
| DrivAerML Runs 1-3 | ~0.79 | ~0.67 | **0.85x (slower)** |
| AirfRANS Run 4 | ~0.44 | ~0.46 | **1.06x** |

**Surprising finding:** No-Fourier DrivAerML runs were actually **17% slower** per epoch than the Fourier baseline, not faster. This contradicts the AirfRANS precedent where no-Fourier was ~3x faster (PR #2455). The speed difference may be due to concurrent GPU usage (4 runs sharing bus bandwidth) or DrivAerML's data loading being the bottleneck rather than Fourier computation.

---

### What happened

**Fourier features are essential for both DrivAerML and AirfRANS.** The hypothesis that Fourier was a holdover from AirfRANS is falsified — Fourier is even MORE critical for DrivAerML than AirfRANS:

1. **DrivAerML without Fourier cannot break 6.78%** even at best, vs 4.619% baseline (47% worse). Most configurations flat-lined at 59%+ or diverged entirely.
2. **AirfRANS without Fourier hits 0.005615** vs 0.001236 baseline (4.5x worse), then diverges.
3. **No epoch speedup materialized** on DrivAerML — the Fourier computation overhead is negligible relative to DrivAerML's data pipeline.
4. **Instability pattern:** All no-Fourier runs showed gradient norm explosion (Infinity or NaN) — Fourier features appear to provide crucial gradient stability by encoding spatial frequencies the model would otherwise need to learn from scratch.

### Suggested follow-ups

- **Fourier is confirmed essential** — all future DrivAerML experiments should continue using `--enable-fourier`.
- The gc=1.0 + WD=1e-2 + lr=7e-4 compound (Run 3) briefly reached 6.78%, suggesting these regularization additions have potential **when combined with Fourier** — could be worth testing as a compound on the Fourier baseline.
- Investigating Fourier frequency bands or learnable Fourier features could potentially improve over the current fixed encoding while maintaining the stability benefits.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2915 because Closing PR #2915 as a confirmed negative result — this is valuable science that definitively establishes Fourier features as non-negotiable for both DrivAerML and AirfRANS.

Key findings:
1. **DrivAerML without Fourier**: Cannot break 6.78% even at best (vs current best 3.997%). Run 1 flat-lined at 59.6% with gradient norms at Infinity. Runs 2 and 3 diverged. The frequency-domain encoding is essential for 3D surface pressure resolution.
2. **AirfRANS without Fourier**: Best was 0.005615 (4.5x worse than 0.000627 current best), then diverged to NaN at ep168.
3. **No epoch speedup on DrivAerML**: No-Fourier runs were 17% *slower* per epoch — the data pipeline is the bottleneck, not Fourier computation. The premise of this ablation was falsified.

**Confirmed directive: all future experiments MUST use --enable-fourier.** This is a hard architectural requirement, not a tunable hyperparameter.

Thank you for rigorous ablation work, levi. Closing with full confidence..
